#162685658 Enable a user to register on the application.

BarnaTB opened the pull request:
**What does this PR do?**
This PR finishes a feature that enables a user to create an account on the application.

**Description of Task to be completed?**
The application should enable a user to register in the application and later use the credentials to log in.

**What are the relevant Pivotal Tracker stories?**
[#162685658](https://www.pivotaltracker.com/story/show/162685658)

**How should it be manually tested?**
Run each of the following commands in your terminal in the respective order.

- `git clone https://github.com/andela/Ah-backend-guardians.git`
- `cd Ah-backend-guardians`
- `git checkout ch-upgrade-162685649`
- `virtualenv venv`
- `. venv/bin/activate`
- The application is configured to use a postgreSQL database so you need to follow the next steps to set up yours so that you can have the feature up and running.
- Create a database `psql -c "create database ahdatabase;" -U postgres`(assuming `postgres` is your admin database user.)
- Create a `.env` file on the root directory and add the data below, substituting each variable as appropriately described.

```
    SECRET_KEY=your-secret-key-here(this is got from your settings.py file)
    DB_NAME=your-database-name-here
    DB_USER=your-database-user-here
    DB_PASSWORD=your-database-password-here
    DB_HOST=your-host-here(most preferrably localhost or 127.0.0.1)
```

- Add the following variables to the same .env file as is.

```
    ALLOWED_HOSTS=.localhost, .herokuapp.com
    EMAIL_HOST=email_host
    EMAIL_HOST_USER=email_host_user
    EMAIL_HOST_PASSWORD=email_host_password
    EMAIL_PORT=EMAIL_PORT
```

- Run migrations `python manage.py makemigrations` then `python manage.py migrate`
- Run the app `python manage.py runserver`
- Hit this url to access the registration endpoint. http://localhost:8000/api/users/
- Copy this data model and paste it into the body of your Postman or any REST test client you are using.

```
{
    "username": "tumuhairw",
    "email": "tumuhairw@mail.com",
    "password": "barnatheuser"
}
```

You should be able to get back the output as in the screenshot.

**Screenshot**
![image](https://user-images.githubusercontent.com/38134382/50971365-e9e22b00-14f4-11e9-8630-d964d7cb2996.png)

archbishop12 commented on authors/apps/authentication/models.py:
```diff
@@ -117,4 +117,20 @@ def get_short_name(self):
+        """This method generates a token"""
+        pass
+
+        # token_expiry = datetime.now() + timedelta(days=60)
```
@BarnaTB Remove commented out code from this pr.

archbishop12 commented:
@BarnaTB Run pep8 command on the files you worked on and fix the violations.

archbishop12 commented on authors/apps/authentication/tests/base.py:
```diff
@@ -0,0 +1,22 @@
+from ..serializers import RegistrationSerializer
+
+
+class BaseTest(APITestCase):
```
@BarnaTB Please rebase from the current develop so as you get the current changes and maintain how the tests were written.

BarnaTB replied:
Done!

archbishop12 commented on authors/apps/authentication/views.py:
```diff
@@ -26,7 +29,20 @@ def post(self, request):
         serializer.save()
 
-        return Response(serializer.data, status=status.HTTP_201_CREATED)
+        token_expiry = datetime.now() + timedelta(days=60)
```
@BarnaTB How did u come up with this time for the token expiry?

BarnaTB replied:
I figured it would not be ideal to always ask the user to login so I made the expiry of their token 60 days, upon which they would not be able to carry out any operations until they log in.
Technically, I get the current time at which the user is registering and add 60 days using the `datetime` and `timedelta` classes from Python's `datetime` module respectively.

archbishop12 replied:
Thanks for clarifying why you chose 60 days.

archbishop12 commented on authors/apps/authentication/views.py:
```diff
@@ -26,7 +29,20 @@ def post(self, request):
+            key=settings.SECRET_KEY,
+            algorithm='HS256')
+        
+        # serializer.data['token'] = token
```
@BarnaTB Remove this if not needed.

archbishop12 commented on authors/apps/authentication/views.py:
```diff
@@ -17,7 +20,7 @@ class RegistrationAPIView(APIView):
-        user = request.data.get('user', {})
+        user = request.data
 
         # The create serializer, validate serializer, save serializer pattern
```
@BarnaTB Update the comments here to what the method actually does and place this on top of the method.

archbishop12 commented on authors/apps/authentication/views.py:
```diff
@@ -26,7 +29,20 @@ def post(self, request):
+        # serializer.data['token'] = token
+        user_token = token.decode('utf-8')
+
+        new_data={'username': user['username'], 'token': user_token}
```
@BarnaTB You change give this variable a more descriptive name  `response_data`.
```suggestion
        response_data={'username': user['username'], 'token': user_token}
```

BarnaTB replied:
Noted. Will do that ASAP. Thanks

archbishop12 commented on authors/apps/authentication/models.py:
```diff
@@ -117,4 +117,20 @@ def get_short_name(self):
         return self.username
 
-
+    def generate_token(self):
```
@BarnaTB Is it possible to move the code for generation of the token here and this method will be called wherever it is needed.

BarnaTB replied:
I had tried that but that also requires me to add the token as a property of this User model and that it hard for me to add it to the `serializer.data` so as to return one response with both the user credentials and token. I will more into it though.

archbishop12 commented on authors/apps/authentication/views.py:
```diff
@@ -17,7 +20,7 @@ class RegistrationAPIView(APIView):
 
     def post(self, request):
-        user = request.data.get('user', {})
+        user = request.data
```
@BarnaTB This changes the type of json object sent, was this agreed upon by the team.

BarnaTB replied:
No. We've not had the discussion yet. I will raise it up to the rest of the team on Monday.

tomuhenry reviewed (approved):
LGTM... The PR does what it's set out to do.

moseskamira committed:
#162685661 Implement a Password Reset functionality (#16)

Feature(PasswordReset) Implement a Password Reset functionality

- write a code that enables a user to reset password and receive a confirmation email

[Finishes #162685661]

BarnaTB committed:
Fix user object conflicts

archbishop12 commented:
@BarnaTB No Token is returned when user signs in.

BarnaTB commented:
@archbishop12 my task only requires me to have the token returned when the user **registers** not login.

BarnaTB commented:
My bad. I've read the description again and it talks about login as well so let me add that right now. Thanks.

archbishop12 commented:
@BarnaTB Here is a screenshot on the acceptance criteria on the PT board.
<img width="480" alt="screen shot 2019-01-15 at 12 52 38" src="https://user-images.githubusercontent.com/39958516/51172534-7a898400-18c4-11e9-900d-7e18e6dc1c81.png">.

archbishop12 reviewed (approved):
@BarnaTB I am merging this
